### DISCUSSION TITLE
Scale JPEG with maximum file size?

### DISCUSSION THREAD

**Discussion by jramos** [libvips/libvips#3815] [Q&A] [answered]
I would like to save a JPEG and have the resulting file be no more than 1MB. With imagemagick's `convert`, I can use `jpeg:extent=1MB` as an argument. From their [docs](https://www.imagemagick.org/Usage/formats/#jpg_write):

>[This] works by generating many versions of the JPEG image, doing a binary search, of the output quality "-quality" setting, until it gets as close as possible to the file size given without exceeding it. It does this by writing the image repeatably into a temporary file and once it has the appropriate quality size, it then outputs the final image to the given output filename, once.

Is there anything comparable in libvips?

**Comment by jcupitt** [comment, marked as answer]
Hi @jramos,

Not directly, but it's easy to make something like that yourself, eg.:

```python
#!/usr/bin/env python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1])

lower = 0
upper = 100
target = 100000
while upper - lower > 5:
    q = lower + (upper - lower) // 2
    data = image.write_to_buffer(".jpg", Q=q)
    print(f"q = {q}, size = {len(data)}")
    if len(data) > target:
        upper = q
    else:
        lower = q

with open(sys.argv[2], "wb") as f:
    f.write(data)

print(f"selected Q = {lower}, written to {sys.argv[2]}")
```

**Reply by kmartinez** [reply]
what I love about John's code here is it shows vips making a series of attempts from the source image on disk to a buffer image type (ie in RAM) until it find the right level - then saves the buffer image. The source image file will be cached by the operating system so after the first pipeline makes the first jpg it will run very fast afterwards..
